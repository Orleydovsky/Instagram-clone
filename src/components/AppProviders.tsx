import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
const queryClient = new QueryClient()

export function AppProviders ({ children }: any) {
  return (
    <React.StrictMode>
      <QueryClientProvider client={queryClient}>
          <Router>{children}</Router>
      </QueryClientProvider>
    </React.StrictMode>
  )
}
