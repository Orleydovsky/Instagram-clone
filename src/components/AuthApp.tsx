import { signOut } from 'firebase/auth'
import React from 'react'
import { auth } from '../services/firebase/firebase-config'

function AuthApp () {
  const logout = async () => {
    await signOut(auth)
  }
  return (
    <div>
        AuthApp
        <button onClick={logout}>Logout</button>
    </div>
  )
}

export default AuthApp
