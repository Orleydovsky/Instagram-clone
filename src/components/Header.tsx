import { signOut } from 'firebase/auth'
import React from 'react'
import { auth } from '../services/firebase/firebase-config'
import instagram from '../assets/instagram.svg'
import { Link } from 'react-router-dom'

export default function Header () {
  return (
    <header className='h-16 bg-white border border-gray-200 mb-5 px-5'>
      <div className='flex justify-between container mx-auto h-full max-w-screen-md items-center'>
        <Link to='/'>
          <img src={instagram} alt='Instagram logo' className='h-8'/>
        </Link>
        <input type='search' placeholder='Search' className='bg-gray-100 rounded-lg py-2 px-3' />
        {
        auth.currentUser
          ? <div className='flex flex-row'>
              <div onClick={async () => await signOut(auth)} className='rounded-full w-7 h-7 bg-black'></div>
            </div>
          : <div className='flex flex-row'>
              <Link to='/' className='bg-blue-500 text-white rounded-md py-1 px-3 font-medium'>Log In</Link>
              <Link to='/signup' className='text-blue-500 py-1 px-3 font-medium'>Sign Up</Link>
            </div>
      }
      </div>
    </header>
  )
}
