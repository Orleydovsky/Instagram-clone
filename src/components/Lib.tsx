import camera from '../assets/camera.svg'
import loading from '../assets/loading.svg'
import React from 'react'
import { Link } from 'react-router-dom'

interface InputProps {
  name: string,
  type: 'text' | 'password'
  placeholder: string,
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void
  value?: string
}

function Input ({ name, type, placeholder, onChange, value }: InputProps) {
  return (
    <input
      onChange={onChange}
      value={value}
      name={name} type={type}
      placeholder={placeholder}
      className='bg-gray-50 border w-full border-gray-200 rounded-sm p-2 text-sm my-1 box-border'
    />
  )
}

interface AvatarProps {
  picture: string | undefined
  size: 'small' | 'medium' | 'large'
}
function Avatar ({ picture, size }: AvatarProps) {
  let applySize
  switch (size) {
    case 'small': applySize = 8; break
    case 'medium': applySize = 16; break
    case 'large': applySize = 40; break
    default: applySize = 8
  }
  return (
    <div className={`w-${applySize} h-${applySize}`}>
      <img src={picture} className='rounded-full aspect-square object-cover w-full h-full'/>
    </div>
  )
}

function NoPostsYet () {
  return (
    <div className='flex flex-col justify-center text-center'>
      <img className='opacity-30 h-8 my-10' src={camera}/>
      <h1 className='text-3xl font-light'>No Posts Yet</h1>
    </div>
  )
}

function NotFollowing () {
  return (
    <div className='col-span-3 text-center lg:col-span-2'>
      <h1 className='font-bold text-xl'>Welcome to Instagram</h1>
      <p>Follow people to start seeing the photos and videos they share</p>
    </div>
  )
}

function NoUserFound () {
  return (
    <div className='text-center container max-w-md mx-auto'>
    <h1 className='font-bold text-xl'>No user found</h1>
    <p>The link you followed may be broken, or the page may have been removed. Go back to <Link className='text-blue-600' to='/'>Instagram.</Link></p>
    </div>
  )
}

function Spinner () {
  return <img src={loading} className='w-5 h-5 animate-spin-slow'/>
}
function FullPageSpinner () {
  return (
    <div className='w-full'>
      <img src={loading} className='w-10 h-10 animate-spin-slow mx-auto'/>
    </div>
  )
}

export {
  Input,
  Avatar,
  NoPostsYet,
  NotFollowing,
  NoUserFound,
  Spinner,
  FullPageSpinner
}
