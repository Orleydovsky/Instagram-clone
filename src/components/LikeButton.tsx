import React from 'react'
import { doc, setDoc, arrayRemove, arrayUnion } from 'firebase/firestore'
import like from '../assets/like.svg'
import liked from '../assets/liked.svg'
import { auth, db } from '../services/firebase/firebase-config'

interface LikeButtonType {
  isLiked: boolean,
  setIsLiked: Function
  postId: string
}

export default function LikeButton ({ isLiked, setIsLiked, postId }: LikeButtonType) {
  const currentUser = auth.currentUser?.uid
  const handleLike = async () => {
    setIsLiked(!isLiked)
    await setDoc(doc(db, 'posts', postId), {
      likes: isLiked ? arrayRemove(currentUser) : arrayUnion(currentUser)
    }, { merge: true })
  }
  return (
    <button className='hover:opacity-50' onClick={handleLike}>
      <img src={isLiked ? liked : like}/>
    </button>
  )
}
