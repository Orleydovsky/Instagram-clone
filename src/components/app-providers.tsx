import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'

interface Props {
  children: JSX.Element
}

export function AppProviders ({ children }: Props) {
  return (
    <React.StrictMode>
      <Router>{children}</Router>
    </React.StrictMode>
  )
}
