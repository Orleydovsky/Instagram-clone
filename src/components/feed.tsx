import { collection, query, where } from 'firebase/firestore'
import { useCollectionData } from 'react-firebase-hooks/firestore'
import { db } from '../services/firebase/firebase-config'
import { FullPageSpinner, NoPostsYet } from './lib'
import Post from './post'

interface Props {
  following: string[],
  uid: string
}

export default function Feed ({ following, uid }: Props) {
  const [posts, isLoading] =
  useCollectionData(
    query(collection(db, 'posts'),
      where('authorId', 'in', [...following, uid])
    )
  )
  return (
    !posts?.length
      ? <div className='grid col-span-3 lg:justify-center lg:items-start justify-center lg:col-span-2 w-full'>
          {isLoading ? <FullPageSpinner/> : <NoPostsYet/>}
        </div>
      : <div className='grid col-span-3 lg:justify-end justify-center lg:col-span-2 w-full'>
          {posts?.map((post) => {
            return <Post key={post.postId} postData={post}/>
          })}
        </div>
  )
}
