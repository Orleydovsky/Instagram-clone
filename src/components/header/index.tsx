import { Link } from 'react-router-dom'
import instagram from '../../assets/instagram.svg'
import React from 'react'
import { HeaderMenuButton } from './menu-button'
import { SearchBar } from './seach-bar'

export default function Header () {
  return (
    <nav className='h-16 bg-white border border-gray-200 px-5 sticky top-0 mb-5 w-full'>
      <div className='flex justify-between container mx-auto h-full max-w-screen-md items-center'>
        <Link to='/'>
          <img src={instagram} alt='Instagram logo' className='h-8'/>
        </Link>
        <SearchBar/>
        <HeaderMenuButton/>
      </div>
    </nav>
  )
}
