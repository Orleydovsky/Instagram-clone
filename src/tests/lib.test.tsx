import renderer from 'react-test-renderer'
import { it, expect, vi } from 'vitest'
import { Avatar, Input, NoPostsYet, NotFollowing, NoUserFound, Spinner, FullPageSpinner } from '../components/lib'
import { BrowserRouter as Router } from 'react-router-dom'

it('renders Input', () => {
  const onChange = vi.fn
  const tree = renderer
    .create(<Input name='name' placeholder='placeholder' type={'text' || 'password'} onChange={onChange}/>)
    .toJSON()
  expect(tree).matchSnapshot()
})
it('renders Avatar', () => {
  const tree = renderer
    .create(<Avatar picture='src' size={'small' || 'medium' || 'large'}/>)
    .toJSON()
  expect(tree).matchSnapshot()
})
it('renders NoPostsYet', () => {
  const tree = renderer
    .create(<NoPostsYet/>)
    .toJSON()
  expect(tree).matchSnapshot()
})
it('renders NotFollowing', () => {
  const tree = renderer
    .create(<NotFollowing/>)
    .toJSON()
  expect(tree).matchSnapshot()
})
it('renders NoUserFound', () => {
  const tree = renderer
    .create(
      <Router>
        <NoUserFound/>
      </Router>
    )
    .toJSON()
  expect(tree).matchSnapshot()
})
it('renders Spinner', () => {
  const tree = renderer
    .create(<Spinner/>)
    .toJSON()
  expect(tree).matchSnapshot()
})
it('renders FullPageSpinner', () => {
  const tree = renderer
    .create(<FullPageSpinner/>)
    .toJSON()
  expect(tree).matchSnapshot()
})
