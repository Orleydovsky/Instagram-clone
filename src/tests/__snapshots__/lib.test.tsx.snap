// Vitest Snapshot v1

exports[`renders Avatar 1`] = `
<div
  className="w-8 h-8"
>
  <img
    className="rounded-full aspect-square object-cover w-full h-full"
    src="src"
  />
</div>
`;

exports[`renders FullPageSpinner 1`] = `
<div
  className="w-full"
>
  <img
    className="w-10 h-10 animate-spin-slow mx-auto"
    src="/src/assets/loading.svg"
  />
</div>
`;

exports[`renders Input 1`] = `
<input
  className="bg-gray-50 border w-full border-gray-200 rounded-sm p-2 text-sm my-1 box-border"
  name="name"
  onChange={[Function]}
  placeholder="placeholder"
  type="text"
/>
`;

exports[`renders NoPostsYet 1`] = `
<div
  className="flex flex-col justify-center text-center"
>
  <img
    className="opacity-30 h-8 my-10"
    src="/src/assets/camera.svg"
  />
  <h1
    className="text-3xl font-light"
  >
    No Posts Yet
  </h1>
</div>
`;

exports[`renders NoUserFound 1`] = `
<div
  className="text-center container max-w-md mx-auto"
>
  <h1
    className="font-bold text-xl"
  >
    No user found
  </h1>
  <p>
    The link you followed may be broken, or the page may have been removed. Go back to 
    <a
      className="text-blue-600"
      href="/"
      onClick={[Function]}
    >
      Instagram.
    </a>
  </p>
</div>
`;

exports[`renders NotFollowing 1`] = `
<div
  className="col-span-3 text-center lg:col-span-2"
>
  <h1
    className="font-bold text-xl"
  >
    Welcome to Instagram
  </h1>
  <p>
    Follow people to start seeing the photos and videos they share
  </p>
</div>
`;

exports[`renders Spinner 1`] = `
<img
  className="w-5 h-5 animate-spin-slow"
  src="/src/assets/loading.svg"
/>
`;
